etna/transforms/math/sklearn.py

```diff
@@ -178,60 +195,89 @@ def inverse_transform(self, df: pd.DataFrame) -> pd.DataFrame:
-                        columns={"target": quantile_column_nm}
-                    )
-
-            else:
-                raise ValueError(f"'{self.mode}' is not a valid TransformMode.")
-            df.loc[:, pd.IndexSlice[:, self.in_column]] = transformed
+            # quantiles inverse transformation
+            for quantile_column_nm in quantiles:
+                df_slice_copy = df.loc[:, pd.IndexSlice[:, self.in_column]].copy()
+                df_slice_copy = set_columns_wide(
+                    df_slice_copy, df, features_left=["target"], features_right=[quantile_column_nm]
+                )
+                transformed_quantile = self._make_inverse_transform(df_slice_copy)
+                df_slice_copy.loc[:, pd.IndexSlice[:, self.in_column]] = transformed_quantile
+                quantiles_arrays[quantile_column_nm] = df_slice_copy.loc[:, pd.IndexSlice[:, "target"]].rename(
+                    columns={"target": quantile_column_nm}
+                )
 
+            df.loc[:, pd.IndexSlice[:, self.in_column]] = transformed
             for quantile_column_nm in quantiles:
                 df.loc[:, pd.IndexSlice[:, quantile_column_nm]] = quantiles_arrays[quantile_column_nm].values
+
         return df
 
-    def _reshape(self, df: pd.DataFrame) -> np.ndarray:
+    def _preprocess_macro(self, df: pd.DataFrame) -> np.ndarray:
         segments = sorted(set(df.columns.get_level_values("segment")))
         x = df.loc[:, pd.IndexSlice[:, self.in_column]]
         x = pd.concat([x[segment] for segment in segments]).values
         return x
 
-    def _inverse_reshape(self, df: pd.DataFrame, transformed: np.ndarray) -> np.ndarray:
+    def _postprocess_macro(self, df: pd.DataFrame, transformed: np.ndarray) -> np.ndarray:
         time_period_len = len(df)
         n_segments = len(set(df.columns.get_level_values("segment")))
         transformed = np.concatenate(
             [transformed[i * time_period_len : (i + 1) * time_period_len, :] for i in range(n_segments)], axis=1
         )
         return transformed
+
+    def _preprocess_per_segment(self, df: pd.DataFrame) -> np.ndarray:
+        self._fit_segments = cast(List[str], self._fit_segments)
+        transform_segments = df.columns.get_level_values("segment").unique()
+        new_segments = set(transform_segments) - set(self._fit_segments)
+        if len(new_segments) > 0:
+            raise NotImplementedError(
+                f"This transform can't process segments that weren't present on train data: {reprlib.repr(new_segments)}"
+            )
+
+        df = df.loc[:, pd.IndexSlice[:, self.in_column]]
+        to_add_segments = set(self._fit_segments) - set(transform_segments)
+        df_to_add = pd.DataFrame(index=df.index, columns=pd.MultiIndex.from_product([to_add_segments, self.in_column]))
+        df = pd.concat([df, df_to_add], axis=1)
+        df = df.sort_index(axis=1)
```

Comment:
Can we do here `df[:, pd.IndexSlice[self._fit_segments, self.in_column]]` just to speed up in case of large ammount of segments ?
It seems like this is a bottleneck, other parts of this method should run in linear time.

Reply:
I see, lets use sorting.

P.S. Tested some alternatives, seems like they using sorting in MultiIndex anyways.